### PR TITLE
planner, executor: fix tp.Flen size when union with castIntAsString

### LINKED ISSUE #8231: result of union length is wrong
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
``` sql
CREATE TABLE t1 (uid int(1));
INSERT INTO t1 SELECT 150;
SELECT 'a' UNION SELECT uid FROM t1;
```
2. What did you expect to see?

``` sql
mysql> SELECT 'a' UNION SELECT uid FROM t1;
+------+
| a    |
+------+
| a    |
| 150  |
+------+
2 rows in set (0.00 sec)
```

3. What did you see instead?

``` sql
tidb> SELECT 'a' UNION SELECT uid FROM t1;
+------+
| a    |
+------+
| 15   |
| a    |
+------+
2 rows in set (0.00 sec)
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
tidb> select tidb_Version();
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_Version()                                                                                                                                                                                                                                                                                                                     |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v2.1.0-rc.3-158-g1082d5f1a
Git Commit Hash: 1082d5f1ac90c7d73b34061ccef7c5cae1b69b6c
Git Branch: master
UTC Build Time: 2018-11-08 07:01:33
GoVersion: go version go1.11 darwin/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.01 sec)
```

### PR DESCRIPTION
**Pull request by AndrewDi**
### What problem does this PR solve?

fix #8231, fix tp.Flen size when union with castIntAsString 

### What is changed and how it works?
When execute `func unionJoinFieldType(a, b *types.FieldType) *types.FieldType`, if any union type is int and result.Flen is less than mysql.MaxIntWidth then set `resultTp.Flen = mysql.MaxIntWidth`

### Check List 

Tests 

 - Unit test

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8442)

### CODE REVIEW

**Comment by shenli** [comment]
@AndrewDi Thanks! Please fix the CI.

**Comment by AndrewDi** [comment]
@shenli it seems like ci server can not download image.
```
Build-agent version 0.1.1073-1f69f340 (2018-11-20T18:07:03+0000)
Starting container golang:1.11
  image cache not found on this host, downloading golang:1.11

Error response from daemon: error parsing HTTP 404 response body: invalid character 'p' after top-level value: "404 page not found\n"
```

**Comment by zhouqiang-cl** [comment]
@AndrewDi I have rerun ci and env is ok. But it failed.  PTAL

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -659,6 +659,9 @@ func unionJoinFieldType(a, b *types.FieldType) *types.FieldType {
 	resultTp.Decimal = mathutil.Max(a.Decimal, b.Decimal)
 	// `Flen - Decimal` is the fraction before '.'
 	resultTp.Flen = mathutil.Max(a.Flen-a.Decimal, b.Flen-b.Decimal) + resultTp.Decimal
+	if (a.EvalType() == types.ETInt || b.EvalType() == types.ETInt) && resultTp.Flen < mysql.MaxIntWidth {
```
Comment (XuHuaiyu) [inline planner/core/logical_plan_builder.go]: When resultTp is also EvalInt, the truncation would not happen. (Check `TestIndexRangeElimininatedProjection`)

We may need to check 
``` go
resultTp.EvalType() != types.ETInt && (a.EvalType() == types.ETInt || b.EvalType() == types.ETInt) && resultTp.Flen < mysql.MaxIntWidth
```

Comment (AndrewDi) [reply planner/core/logical_plan_builder.go]: Good, check more strict.

**Comment by AndrewDi** [comment]
> @AndrewDi I have rerun ci and env is ok. But it failed. PTAL

That's wired, I just update branch, and CI check successful.

**Comment by zz-jason** [comment]
@XuHuaiyu PTAL

**Comment by zz-jason** [comment]
@AndrewDi Could you cherry-pick this commit to the release-2.1 and release-2.0 branch?

**Pull request merged**